### PR TITLE
OTA(Over-The-Air) Upgrades For Edge Node

### LINKED ISSUE #5589: [OSPP] OTA(Over-The-Air) Upgrades For Edge Node
**What would you like to be added/modified**:

- [ ] Update the proposal of the NodeUpgradeJob;
- [ ] During the edge node upgrade, the keadm can verify the image digest. If the image is invalid,  the edge node cannot be upgraded and an error message is reported;
- [ ] If the edge node upgrade confirmation is enabled, the upgrade cannot be executes before the confirmation and the job will always be waiting;
- [ ] Users can call the metaserver API or execute command `keadm ctl ...`  to confirm the upgrade;

**Why is this needed**:

In order to make the edge node more convenient and rapid upgrade, we introduce a remote upgrade scheme OTA (Over-The-Air) into KubeEdge. In the main process of OTA(i.e. make the bundle, download the bundle, verify the bundle and firmware upgrade), we have realized most steps. Our release will generate a new image version called installation-package, then we use the NodeUpgradeJob CRD to obtain the installation tool keadm in the image and run the command to upgrade the edge node. During this process, if the hacker masquerades the image in the edge node, this will result in the untrusted binary keadm. We need to verify the digest of the image before the keadm executes the upgrade, which is the third step of OTA to verify the bundle. And in some business scenarios (Internet of vehicles, Internet of Things), we also need to provide an option to make the node wait for confirmation from a person with permission before upgrading the edge node. In order to fulfill the above requirements, we need you to complete these tasks:
1. We expect you to complete the validation of the image digest before the edge node upgrade;
2. We expect you to add a field to define whether the edge node upgrade confirmation is required. If required, wait for the confirmation before upgrade the node;
3. We expect you to provide an API in MetaService to confirm the edge node upgrade, and provide a command in the subcommand `keadm ctl` too;

**Refer**:

- https://github.com/kubeedge/kubeedge/blob/master/docs/proposals/edge-node-tasks-design.md

### PR DESCRIPTION
**Pull request by HT0403**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
In order to make the edge node more convenient and rapid upgrade, we introduce a remote upgrade scheme OTA (Over-The-Air) into KubeEdge. In the main process of OTA(i.e. make the bundle, download the bundle, verify the bundle and firmware upgrade), we have realized most steps. Our release will generate a new image version called installation-package, then we use the NodeUpgradeJob CRD to obtain the installation tool keadm in the image and run the command to upgrade the edge node. During this process, if the hacker masquerades the image in the edge node, this will result in the untrusted binary keadm. We need to verify the digest of the image before the keadm executes the upgrade, which is the third step of OTA to verify the bundle. And in some business scenarios (Internet of vehicles, Internet of Things), we also need to provide an option to make the node wait for confirmation from a person with permission before upgrading the edge node.
**Which issue(s) this PR fixes**:

Fixes #5589 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cloud/pkg/taskmanager/nodeupgradecontroller/node_upgrade_controller.go [hunk 1]
```diff
@@ -47,6 +41,16 @@ import (
...
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachineryType "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog/v2"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/retry"
```
Comment (wbc6080) [inline cloud/pkg/taskmanager/nodeupgradecontroller/node_upgrade_controller.go:53]: Need sort the import. Packages starting with `github.com/kubeedge` should be placed in the third part

File: edge/pkg/edgehub/task/taskexecutor/node_upgrade.go [hunk 2]
```diff
@@ -161,6 +170,19 @@ func prepareKeadm(upgradeReq *commontypes.NodeUpgradeJobRequest) error {
...
+	if upgradeReq.ImageDigest != "" {
+		var local string
+		local, err = container.GetImageDigest(image)
+		if err != nil {
+			return err
+		}
+		// TODO: get local installation-package digest
+		if upgradeReq.ImageDigest != local {
+			err = fmt.Errorf("invalid installation-package image digest value: %s", local)
+			return err
```
Comment (wbc6080) [inline edge/pkg/edgehub/task/taskexecutor/node_upgrade.go:183]:
```suggestion
			return fmt.Errorf("invalid installation-package image digest value: %s", local)
```

File: edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go [hunk 3]
```diff
@@ -1,10 +1,22 @@
...
 	"net/http"
+	"os/exec"
+	"reflect"
+	"strconv"
 
+	"github.com/beego/beego/v2/client/orm"
+	"github.com/kubeedge/kubeedge/common/types"
+	commontypes "github.com/kubeedge/kubeedge/common/types"
+	"github.com/kubeedge/kubeedge/edge/pkg/edgehub/task/taskexecutor"
+	v2 "github.com/kubeedge/kubeedge/edge/pkg/metamanager/dao/v2"
```
Comment (wbc6080) [inline edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:16]: ditto

**Review by wbc6080 — COMMENTED** [review]
Please make the verify CI check pass. You can refer to https://kubeedge.io/docs/developer/make_verify

**Comment by wbc6080** [comment]
Please fill in the **What this PR does / why we need it** field in detail.
image

**Comment by HT0403** [comment]
I have run `make_verify`,but it is still failed. @wbc6080

**Review by wbc6080 — COMMENTED** [review]
Follow the prompts to check the corresponding file code normative issues.
image

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
+package upgradedb
```
Comment (Shelley-BaoYue) [inline edge/pkg/metamanager/dao/upgradedb/confirm_db.go:1]: As we have discussed before, we decided save confirm data in metav2 but not create a new table, right?

File: keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	if statusCode != 200 {
+		return nil, fmt.Errorf("node upgrade confirm failed with status code: %d", statusCode)
+	}
+
+	body, err := result.Raw()
+	if err != nil {
+		return nil, err
+	}
+
+	var restartResponse types.RestartResponse
```
[the remark this answers is not in the document]
Comment (HT0403) [reply keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go:94]: I will add a struct called `NodeUpgradeConfirmResponse`.

File: keadm/cmd/keadm/app/cmd/common/constant.go [hunk 6]
```diff
@@ -50,6 +50,9 @@ const (
 
 	// FlagNamePrintFinalValues ...
 	FlagNamePrintFinalValues = "print-final-values"
+
+	// FlagNameUpgradeJobName
+	FlagNameUpgradeJobName = "node-upgrade-name"
```
[the remark this answers is not in the document]
Comment (HT0403) [reply keadm/cmd/keadm/app/cmd/common/constant.go:55]: This constant is not used and I plan to delete it.

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go (hunk 4 above)
Comment (HT0403) [reply edge/pkg/metamanager/dao/upgradedb/confirm_db.go:1]: I created a table called `NodeUpgradeConfirmTable` and registered this table using `orm.RegisterModel`,and then I use database to save `NodeTaskRequest` and `NodeUpgradeJobRequest`.

File: hack/update-codegen.sh [hunk 7]
```diff
@@ -32,9 +32,12 @@ cleanup
...
+
 
-${KUBEEDGE_ROOT}/hack/generate-groups.sh "deepcopy,client,informer,lister" \
-github.com/kubeedge/api/client github.com/kubeedge/api/apis \
-"devices:v1beta1 reliablesyncs:v1alpha1 rules:v1 apps:v1alpha1 operations:v1alpha1 policy:v1alpha1" \
---go-header-file ${KUBEEDGE_ROOT}/hack/boilerplate/boilerplate.txt
+"${KUBEEDGE_ROOT}/hack/generate-groups.sh" "deepcopy,client,informer,lister" \
+  github.com/kubeedge/api/client github.com/kubeedge/api/apis \
+  "devices:v1beta1 reliablesyncs:v1alpha1 rules:v1 apps:v1alpha1 operations:v1alpha1 policy:v1alpha1" \
+  --go-header-file ${KUBEEDGE_ROOT}/hack/boilerplate/boilerplate.txt
```
[the remark this answers is not in the document]
Comment (HT0403) [reply hack/update-codegen.sh:43]: I have restored the modifications, which may have been modified when generating the CRD.

File: common/types/types.go [hunk 8]
```diff
@@ -110,3 +112,8 @@ type RestartResponse struct {
 	ErrMessages []string `json:"errMessages,omitempty"`
 	LogMessages []string `json:"LogMessages,omitempty"`
 }
+
+type NodeUpgradeConfirmResponse struct {
+	ErrMessages []string `json:"errMessages,omitempty"`
+	LogMessages []string `json:"LogMessages,omitempty"`
```
Comment (wbc6080) [inline common/types/types.go:118]:
```suggestion
	LogMessages []string `json:"logMessages,omitempty"`
```

File: edge/pkg/edgehub/task/taskexecutor/node_upgrade.go [hunk 9]
```diff
@@ -79,15 +81,48 @@ func initUpgrade(taskReq types.NodeTaskRequest) (event fsm.Event) {
 	}
 
 	if upgradeReq.UpgradeID == "" {
-		err = fmt.Errorf("upgradeID cannot be empty")
+		err = errors.New("upgradeID cannot be empty")
 		return
 	}
+	if upgradeReq.RequireConfirmation {
+		orm := upgradedb.InitNodeUpgradeConfirmTable()
+		var upgradeJobReqDb = commontypes.NodeUpgradeJobRequest{
```
Comment (wbc6080) [inline edge/pkg/edgehub/task/taskexecutor/node_upgrade.go:89]: Use `upgradeJobReqDB`  may be better?

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+		if err != nil {
+			return commontypes.NodeUpgradeJobRequest{}, err
+		}
+		nodeUpgradeReqValue.FieldByName(data.Name).Set(reflect.ValueOf(data.Value))
+	}
+	return nodeUpgradeReq, nil
+}
+
+// SaveDeviceTwin save NodeTaskRequestField
+func saveNodeUpgradeConfirm(o orm.Ormer, doc *NodeUpgradeConfirmTable) error {
```
Comment (wbc6080) [inline edge/pkg/metamanager/dao/upgradedb/confirm_db.go:71]:
```suggestion
// saveNodeUpgradeConfirm save ...
func saveNodeUpgradeConfirm(o orm.Ormer, doc *NodeUpgradeConfirmTable) error {
```

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+		_, e := txOrm.Insert(doc)
+		// if e != nil the transaction will be rollback
+		// or it will be committed
+		return e
+	})
+	if err != nil {
+		klog.Errorf("Something wrong when insert NodeTaskRequest data: %v", err)
+		return err
+	}
+	klog.V(4).Info("insert NodeTaskRequest data successfully")
```
Comment (wbc6080) [inline edge/pkg/metamanager/dao/upgradedb/confirm_db.go:84]: Is the log prompt information inaccurate? It should be the` NodeUpgrade` task?

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+		// or it will be committed
+		return e
+	})
+
+	if err != nil {
+		klog.Errorf("Something wrong when deleting NodeTaskRequest data: %v", err)
+		return err
+	}
+	klog.V(4).Info("Delete NodeTaskRequest data successfully")
+	return nil
```
Comment (wbc6080) [inline edge/pkg/metamanager/dao/upgradedb/confirm_db.go:142]: Should be `NodeTaskRequest` or `NodeUpgradeConfirm` task?

File: edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go [hunk 13]
```diff
@@ -42,3 +53,35 @@ func (f *Factory) Restart(namespace string) http.Handler {
 	})
 	return h
 }
+func (f *Factory) ConfirmUpgrade() http.Handler {
+	h := http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		klog.Infof("Begin to run upgrade command")
```
Comment (wbc6080) [inline edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:58]:
```suggestion
		klog.Info("Begin to run upgrade command")
```

File: keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+// NewEdgeConfirm returns KubeEdge confirm command.
+func NewEdgeConfirm() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "confirm",
+		Short: edgeConfirmShortDescription,
+		Long:  edgeConfirmShortDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) <= 0 {
+				return fmt.Errorf("no specified node name for confirm upgrade")
```
Comment (wbc6080) [inline keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go:44]: Use `errors.New()`

File: keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+		return err
+	}
+	ctx := context.Background()
+	confirmResponse, err := nodeConfirm(ctx, kubeClient)
+	if err != nil {
+		return err
+	}
+
+	for _, logMsg := range confirmResponse.LogMessages {
+		fmt.Println(logMsg)
```
Comment (wbc6080) [inline keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go:64]: Use klog

File: keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	for _, errMsg := range confirmResponse.ErrMessages {
+		fmt.Println(errMsg)
```
Comment (wbc6080) [inline keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go:68]: ditto

**Review by wbc6080 — COMMENTED** [review]
Please squash some commits https://kubeedge.io/docs/developer/Supplemental_developer_guide#ensuring-to-squash-similiar-commits

File: common/types/types.go (hunk 8 above)
Comment (HT0403) [reply common/types/types.go:118]: OK

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go (hunk 11 above)
Comment (HT0403) [reply edge/pkg/metamanager/dao/upgradedb/confirm_db.go:84]: This method is using `NodeUpgradeConfirmTable` to save `NodeTaskRequest` struct and `NodeUpgradeJobRequest` struct.

File: edge/pkg/metamanager/dao/upgradedb/confirm_db.go (hunk 12 above)
Comment (HT0403) [reply edge/pkg/metamanager/dao/upgradedb/confirm_db.go:142]: `NodeUPgradeConfirm`

File: keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go (hunk 14 above)
Comment (HT0403) [reply keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go:44]: Yes

File: keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go (hunk 15 above)
Comment (HT0403) [reply keadm/cmd/keadm/app/cmd/ctl/confirm/confirm.go:64]: Ok

**Review by fisherxu — COMMENTED** [review]
Is the PR ready for review? Since it still has the `WIP` label :) @HT0403 @WillardHu

**Comment by HT0403** [comment]
This PR is ready for review, I remove the `WIP` label, PTA @fisherxu @wbc6080 @WillardHu

File: edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go [hunk 17]
```diff
@@ -42,3 +52,39 @@ func (f *Factory) Restart(namespace string) http.Handler {
...
+		// use nohup command to start a child progress
+		command := fmt.Sprintf("nohup %s &", upgradeCmd)
+		cmd := exec.Command("bash", "-c", command)
+		s, err := cmd.CombinedOutput()
+		if err != nil {
+			http.Error(w, fmt.Sprintf("run upgrade command %s failed: %v, res: %s", command, err, s),
+				http.StatusInternalServerError)
+			return
+		}
+		klog.Infof("!!! Finish upgrade from Version %s to %s ...", version.Get(), upgradeReq.Version)
```
Comment (fisherxu) [inline edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:79]:
```suggestion
		klog.Infof("Finish upgrade from Version %s to %s ...", version.Get(), upgradeReq.Version)
```

**Review by fisherxu — COMMENTED** [review]
Overall looks good, thanks!

File: edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go (hunk 17 above)
Comment (HT0403) [reply edge/pkg/metamanager/metaserver/handlerfactory/ext_handler.go:79]: OK，thank you for your comment

**Review by fisherxu — COMMENTED** [review]
/approve
Looks good for the first version!

**Pull request merged**